src/main/java/cubicchunks/worldgen/dependency/DependencyManager.java

```diff
@@ -0,0 +1,188 @@
+ *  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ *  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ *  THE SOFTWARE.
+ */
+
+package cubicchunks.worldgen.dependency;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Set;
+
+import cubicchunks.CubicChunks;
+import cubicchunks.server.ServerCubeCache;
+import cubicchunks.server.ServerCubeCache.LoadType;
+import cubicchunks.util.AddressTools;
+import cubicchunks.util.CubeCoords;
+import cubicchunks.world.cube.Cube;
+import cubicchunks.worldgen.GeneratorPipeline;
+
+public class DependencyManager {
+
+	private ServerCubeCache cubeProvider;
+	
+	private GeneratorPipeline generatorPipeline;
+	
+	private Map<CubeCoords, HashSet<Dependent>> requirementsToDependents;
+
+	private Map<CubeCoords, Dependent> dependentMap;
+	
+	
+	public DependencyManager(ServerCubeCache cubeProvider, GeneratorPipeline generatorPipeline) {
+		this.cubeProvider = cubeProvider;
+		this.generatorPipeline = generatorPipeline;
+		this.dependentMap = new HashMap<CubeCoords, Dependent>();
+		this.requirementsToDependents = new HashMap<CubeCoords, HashSet<Dependent>>();
+	}
+	
+	
+	public void initialize(Dependent dependent) {
+		for (Requirement requirement : dependent.requirements.values()) {
+			Cube cube = cubeProvider.getCube(requirement.getCoords());
+			if (cube != null) {
+				dependent.update(this, cube);
+			}
+		}
+	}
+	
+
+	// Updates all dependents waiting for the given cube.
+	public boolean updateDependents(Cube requiredCube) {
+		HashSet<Dependent> dependents = this.requirementsToDependents.get(requiredCube.getCoords());
+		if (dependents != null) {
+			
+			Iterator<Dependent> iter = dependents.iterator();
+			while (iter.hasNext()) {
+				Dependent dependent = iter.next();
+				
+				dependent.update(this, requiredCube);
+				
+				if (dependent.remaining == 0) {
+					generatorPipeline.resume(dependent.cube);
+				}
+			}
+			return true;
+		}
+		return false;
+	}
+	
+	public boolean isSatisfied(Requirement requirement) {
+		return this.cubeProvider.cubeExists(requirement.getCoords());
+	}
+	
+	/**
+	 * Adds a new requirement for the given dependent.
+	 * The requirement must not yet be part of the dependent's map of requirements.
+	 * 
+	 * @param dependent The dependent to which the requirement is to be added.
+	 * @param requirement The requirement to be added to the dependent.
+	 */
+	public void addRequirement(Dependent dependent, Requirement requirement) {
+		
+		// Does the dependent already depend on the cube?
+		Requirement existing = dependent.requirements.get(requirement.getCoords());
+		if (existing != null) {
+			
+			// If it does, return.
+			if (existing.contains(requirement)) {
+				return;
+			}
+			
+			// Otherwise, replace the old requirement.
+			dependent.requirements.get(requirement.getCoords());
```

Comment:
Well, that's a good question o.O Should be
`dependent.requirements.put(requirement.getCoords(), requirement)`